feat(client): login page and google login

hjin03 opened the pull request:
sign in func

- Form Input : 아이디, 비밀번호
- sign-in button
- 아이디 찾기, 비밀번호 찾기, 회원가입 링크 
- Google로 로그인 -> callback page로 redirection

실제 서버와 연결 X - 로그인 버튼 클릭 시 isValid는 현재 false를 반환하여 로그인 오류 문구 표시
현재 구글 로그인시 현재 루트페이지로 연결되도록 설정

hjin03 committed:
feat(client): login page and google login

DonghyunKim98 commented on apps/web/src/app/layout.tsx:
```diff
@@ -25,9 +25,11 @@ export default function RootLayout({
     <html lang='ko' suppressHydrationWarning>
       <body
-        className={`${geistSans.variable} ${geistMono.variable} antialiased`}
+        className={`${geistSans.variable} ${geistMono.variable} antialiased flex justify-center bg-white`}
```
(단순 궁금)
- tailwind 는 런타임에서 동적으로 className 관리를 할 수 없는 것으로 아는데요, 단순한 상수화 과정일까요?

hjin03 replied:
초기 세팅한 분의 말로는 저 variable 값은 빌드 시점에서 결정돼서 괜찮을 거라고 합니다. 말씀대로 단순 상수화를 위한 과정인 것 같습니다!

DonghyunKim98 commented on apps/web/src/app/sign-in/page.tsx:
```diff
@@ -0,0 +1,5 @@
+import { SigninContainer } from '@/containers/sign-in.container';
+
+export default function Page() {
+  return <SigninContainer></SigninContainer>;
```
```suggestion
  return <SigninContainer/>;
```

DonghyunKim98 commented on apps/web/src/containers/sign-in.callback.container.tsx:
```diff
@@ -0,0 +1,23 @@
+  const searchParams = useSearchParams();
+  const router = useRouter();
+
+  useEffect(() => {
+    const code = searchParams.get('code');
+
+    if (code) {
+      // TODO: 받은 code로 백엔드 API 호출하여 처리
+      console.log('인증 코드:', code);
+      // TODO: 처리 완료 후 원하는 페이지로 이동
+      router.push('/');
+    }
+  }, [searchParams]);
```
- [토스 팀에서 제시하는 좋은 코드](https://frontend-fundamentals.com/code/examples/login-start-page.html)
- 위 링크를 읽어보시고, Wrapper 형식 (현재 같은 경우 page.tsx 에서) 처리해주세요
- Server Component 인 Page.tsx 에서 [redirect](https://nextjs.org/docs/app/api-reference/functions/redirect) 를 처리하는 방식은 따로 있습니다 참고해주세요

hjin03 replied:
어떤 부분을 wrapper로 만들어야 할 지 감이 잘 잡히지 않습니다ㅠㅠ 로그인 로직 전체를 wrapper로 만드는걸까요....?

DonghyunKim98 replied:
Container 를 Page.tsx 에서 부르는거 아닌가요? Page.tsx 에서 이 useEffect 부분을 따로 빼는게 맞을 것 같습니다.

DonghyunKim98 commented on apps/web/src/containers/sign-in.container.tsx:
```diff
@@ -0,0 +1,48 @@
+import { SigninPresenter } from './sign-in.presenter';
+
+export function SigninContainer() {
+  const [isValid, setIsValid] = useState(true);
```
- derived State 입니다. react-hook-form 의 `formState.isValid` 로 관리할수 있어요.

hjin03 replied:
해당 isValid는 입력필드의 유효성 검사보다는 실제 서버와 연결했을 때 로그인 성공/실패를 나타내기 위해 사용하였습니다..!

코드에서는 아래와 같이 사용하였습니다.
{isValid ? '' : '아이디 또는 비밀번호가 틀렸습니다!'}

formState의 isValid는 입력값의 유효성 검사를 위한 것으로 알고있는데 맞을까요?

DonghyunKim98 replied:
그런 경우라면 [setError](https://react-hook-form.com/docs/useform/seterror) 를 이용하여, submit 단계에서 일어난 오류를 설정하는 편이 더 나을것 같아요. 지금은 react-hook-form 을 사용하면서 derived state 를 사용하는것으로 보여집니다.

DonghyunKim98 commented on apps/web/src/containers/sign-in.container.tsx:
```diff
@@ -0,0 +1,48 @@
+export function SigninContainer() {
+  const [isValid, setIsValid] = useState(true);
+
+  const formSchema = z.object({
+    id: z.string(),
+    password: z.string(),
+  });
```
- 컴포넌트 안에서 렌더링 될때마다 스키마를 생성할 필요는 없어보여요. 따로 빼주세요.

DonghyunKim98 commented on apps/web/src/containers/sign-in.container.tsx:
```diff
@@ -0,0 +1,48 @@
+    const clientId = process.env.NEXT_PUBLIC_GOOGLE_AUTH_CLIENT_ID;
+    const redirectUri = process.env.NEXT_PUBLIC_OAUTH_REDIRECT;
+
+    const googleAuthUrl = `https://accounts.google.com/o/oauth2/v2/auth?client_id=${clientId}&redirect_uri=${redirectUri}&response_type=code&scope=email%20profile`;
```
- 상수화를 해주세요.

DonghyunKim98 commented on apps/web/src/containers/sign-in.container.tsx:
```diff
@@ -0,0 +1,48 @@
+    window.location.href = googleAuthUrl;
+  };
+
+  function onSubmit() {
+    console.log(form.getValues());
+
+    // TODO: 추후 api 연동 후 로그인 성공/실패 로직 구현
+    setIsValid(false);
+  }
```
- handleSubmit 의 첫번째 인자로 오는 [onSubmit](https://react-hook-form.com/docs/useform/handlesubmit) 에선 첫번째 파라미터로 입력된 데이터가 옵니다

```suggestion
  function onSubmit(data : z.infer<typeof formSchema>) {
   console.log(data);
  }
```

요런식으로 구현해주세요.

DonghyunKim98 commented on apps/web/src/containers/sign-in.presenter.tsx:
```diff
@@ -0,0 +1,96 @@
+  onGoogleLogin: () => void;
+};
+
+export function SigninPresenter({
+  form,
+  onSubmit,
+  isValid,
+  onGoogleLogin,
+}: SigninPresenterProps) {
```
- react-hook-form 의 useFormContext 를 이용해 직접 form 을 받지 말아주세요.
- 자세한 리뷰사항은 [이 링크](https://github.com/Flowery-org/Flowery-client/pull/10#discussion_r1924579117)를 참고해주세요

DonghyunKim98 reviewed (changes requested):
전반적으로 Form 과 관련된 수정이 불가피해보입니다. 지수님과 이야기해서 Form 과 관련된 사항 수정이 먼저 필요할 것으로 보입니다.

hjin03 committed:
feat(client): login feature with jwt and middleware

fix(client): fix login logic and add jwt

chore(client): rename and organize files

chore(client): fix middleware and rename file

DonghyunKim98 commented on apps/web/src/app/sign-in/page.tsx:
```diff
@@ -0,0 +1,21 @@
+    redirect('/');
+  }
+
+  async function handleLogin(accessToken: string, refreshToken: string) {
+    'use server';
+    await setAuthTokens(accessToken, refreshToken);
+  }
```
개인적으로 서버액션은 순수함수인 경우 파일 분리가 나은것 같아요.

DonghyunKim98 commented on apps/web/src/components/common/form-input.tsx:
```diff
@@ -10,24 +10,29 @@ import {
   className,
   ...props
 }: FormInputProps<T>) {
+  const form = useFormContext<T>();
+  const { dirtyFields, errors } = form.formState;
+  const isFieldValid =
+    dirtyFields[name as keyof typeof dirtyFields] && !errors[name];
```
1. name 은 보통 props 로 주입받던가 하드코딩해서 넣어줘도 됩니다. `FormInput` 컴포넌트는 범용 컴포넌안가요?
2. react-hook-form 은 deconstructing 을 권장합니다. 전체적으로 form.formState 같이 사용하지 마시고 `const {formState} = form` 등으로 사용해주세요. (control, handleSubmit 도 마찬가지에요)

hjin03 replied:
이 form-input 컴포넌트 수정 부분은 지수님쪽에서도 함께 쓰이는 컴포넌트라
나중에 merge 할 때 지수님이 수정한 것으로 따라가도록 하겠습니다!

DonghyunKim98 commented on apps/web/src/containers/sign-in.container.tsx:
```diff
@@ -0,0 +1,65 @@
+  });
+
+  const loginWithGoogle = () => {
+    const clientId = process.env.NEXT_PUBLIC_GOOGLE_AUTH_CLIENT_ID;
+    const redirectUri = process.env.NEXT_PUBLIC_OAUTH_REDIRECT;
+    const googleAuthUrl = `${GOOGLE_AUTH_URL}?client_id=${clientId}&redirect_uri=${redirectUri}&response_type=code&scope=email%20profile`;
```
결국 필요한건 googleAuthUrl 인데 나머지 부분들을 지역변수로 할당할 필요는 없는 것 같아요 컴포넌트 밖으로 빼던가 파일로 분리해주세요.

DonghyunKim98 commented on apps/web/src/containers/sign-in.container.tsx:
```diff
@@ -0,0 +1,65 @@
+
+  return (
+    <FormProvider {...form}>
+      <SigninPresenter
+        onSubmit={onSubmit}
+        onGoogleLogin={loginWithGoogle}
+      ></SigninPresenter>
```
```suggestion
      <SigninPresenter
        onSubmit={onSubmit}
        onGoogleLogin={loginWithGoogle}
     />
```

standardwish replied:
이거 말할까 말까 고민했는데 말해주셔서 감사합니다

DonghyunKim98 reviewed (changes requested):
access / refresh 관련 부분은 동작한다고 믿고 넘어가겠습니다. 로직상 중대하게 틀렸다고 보여지는건 없었어요.

standardwish commented on apps/web/src/containers/sign-in.container.tsx:
```diff
@@ -0,0 +1,65 @@
+  async function onSubmit(data: z.infer<typeof SigninFormSchema>) {
+    try {
+      const response = await fetch(
+        `${process.env.NEXT_PUBLIC_API_URL}/auth/login`,
```
서버와 맞출 내용이긴 하지만 기왕이면 용어를 `sign-in`으로 통일하는 게 좋아보입니다

standardwish commented on apps/web/src/containers/sign-in.presenter.tsx:
```diff
@@ -0,0 +1,90 @@
+            <span className='mx-4'>or</span>
+            <div className='w-full h-[0.7px] bg-gray-700' />
+          </div>
+          <Button
```
구글 로그인 버튼 디자인이 변경되었습니다. 피그마 참고해서 수정해주세요!

standardwish replied:
또한 구글 로그인 버튼이 로그인 페이지에서만 사용되므로 컴포넌트화가 안 되어 있는 거 같은데, `SigninPresenter` 함수 내에서라도 별도의 함수로 빼는 건 어떨까요? 코드 보기 이쁘라고 하는 거긴 합니다
```javascript
function GoogleSignInButton() {
    return (<Button
      className='bg-[#EDEAE4] text-black border-gray-300 border hover:bg-gray-600/30'
      variant='secondary'
      size='lg'
      onClick={onGoogleLogin}
    >
      <Image src='Google.svg' alt='Icon' width='24' height='24' />
      구글로 로그인
    </Button>)
  }
...
<div className='py-12 justify-center flex flex-col gap-4'>
          ...
          <GoogleSignInButton />
          ...
</div>
```

standardwish commented on apps/web/src/containers/sign-in.presenter.tsx:
```diff
@@ -0,0 +1,90 @@
+        </div>
+        <span className='text-4xl'>환영합니다!</span>
+      </div>
+      <div className=''>
```
빈 클래스네임 지워주세요

standardwish commented on apps/web/src/containers/sign-in.presenter.tsx:
```diff
@@ -0,0 +1,90 @@
+    formState: { errors },
+  } = form;
+  return (
+    <div className='px-8 '>
```
뒤에 공백 지워주세요

standardwish commented on apps/web/src/containers/sign-in.presenter.tsx:
```diff
@@ -0,0 +1,90 @@
+        <div>
+          <Form {...form}>
+            <form onSubmit={form.handleSubmit(onSubmit)}>
+              <FormInput
```
name 타입 추론을 위해 `<FormInput<FormValues>`로 바꿔보는 건 어떨까요? name이 구분자이다 보니 오타 등의 휴먼 에러에 취약해 보입니다!

standardwish commented on apps/web/src/middleware.ts:
```diff
@@ -0,0 +1,72 @@
+import type { NextRequest } from 'next/server';
+
+const PUBLIC_ROUTES = ['/sign-in', '/sign-up', '/password-reset'];
+const PROTECTED_ROUTES = ['/about', '/profile', '/settings'];
```
통일성을 위해서 두 가지 상수들도 `constants`폴더에 포함시키는 건 어떨까요? `routes.constant.ts` 이런식으로요~~ 현재 `constant` 폴더 내에 있는 google.ts 같은 경우에도 파일 이름에 의해 혼동할 수 있으므로 `google.constant.ts`로 바꾸는 패턴을 사용하면 좋을 거 같습니다

standardwish commented on apps/web/src/service/google-login.service.ts:
```diff
@@ -0,0 +1,49 @@
+
+  try {
+    const googleUserInfo = await getGoogleUserInfo(code);
+    console.log('googleUserInfo : ', googleUserInfo);
```
콘솔 로그 삭제할 것

standardwish commented on apps/web/src/service/google-login.service.ts:
```diff
@@ -0,0 +1,49 @@
+    if (!response.ok) {
+      if (response.status === 404) {
+        redirect(
+          `/sign-up?email=${googleUserInfo.email}&name=${googleUserInfo.name}`,
```
뭔가 url에 이메일과 이름을 노출시키고 싶지 않은데, url params 말고는 따로 방법이 떠오르지 않아 고민되네요

hjin03 replied:
cookie에 저장해서 넘기는 방식은 어떨까요

standardwish commented on apps/web/src/types/sign-in.ts:
`GoogleUserInfo`나 `TokenResponse` 정도를 제외하고는 별도의 타입 파일로 관리할 필요가 없지 않나요? 예를 들어 `SigninPresenterProps`의 경우 `sign-in.presenter.tsx` 파일에서만 사용하는 것으로 보입니다. 또한 파일 명을 sign-in.type.ts로 변경하고 싶은 마음이 듭니다

hjin03 replied:
그럼 한 파일에서만 쓰이는건 전부 파일 내부에서 정의하도록 수정할까요?

standardwish commented on apps/web/src/types/sign-in.ts:
```diff
@@ -0,0 +1,33 @@
+  password: z.string(),
+});
+
+export type FormValues = z.infer<typeof SigninFormSchema>;
```
해당 이름이 혼동을 줄 여지가 매우 커보여서, 명확하게 `SignInFormValues`로 명시하는 게 좋아보입니다.

standardwish commented on apps/web/src/util/fetcher.ts:
LGTM

standardwish commented on apps/web/src/util/google-login.tsx:
파일명을 함수 내용에 알맞게 수정해주세요

standardwish commented on packages/ui/src/styles/globals.css:
원래대로 다른 색깔 형식과 통일할 것

hjin03 replied:
이건 나중에 지수님 merge하는거 따라갈기요

standardwish reviewed (changes requested):
access-token, refresh-token 관련 로직은 큰 문제 없는 것 같고, 피그마 디자인이 바뀌어서 그거 맞게 다시 수정해주세요! 수고하셨습니다.

standardwish commented on apps/web/src/constants/google.ts:
google.constant.ts

standardwish commented on apps/web/src/containers/sign-in.container.tsx:
```diff
@@ -0,0 +1,67 @@
+import { z } from 'zod';
+import { SigninPresenter } from './sign-in.presenter';
+
+export type SigninContainerProps = {
```
SigninContainerProps는 여기서만 쓰니까 export 안해도 될 거 같습니다~

standardwish commented on apps/web/src/containers/sign-in.presenter.tsx:
```diff
@@ -0,0 +1,103 @@
+
+export type SigninFormValues = z.infer<typeof SigninFormSchema>;
+
+export type SigninPresenterProps = {
```
마찬가지로 파일 내에서만 사용하므로 export 제거

standardwish commented on apps/web/src/containers/sign-in.presenter.tsx:
```diff
@@ -0,0 +1,103 @@
+import { useFormContext } from 'react-hook-form';
+import { z } from 'zod';
+
+export type SigninFormValues = z.infer<typeof SigninFormSchema>;
```
마찬가지로 파일 내에서만 사용하므로 export 제거

standardwish commented on apps/web/src/middleware.ts:
```diff
@@ -0,0 +1,70 @@
+}
+
+export const config = {
+  matcher: [...PUBLIC_ROUTES, ...PROTECTED_ROUTES],
```
현재 파일을 실행시키면 다음과 같은 오류가 발생합니다.
```
Unable to parse config export in source filee file need to have a very specific format from which some properties can be statically parsed at compiled-time.
│ The exported configuration object in a source file need to have a very specific format from which some properties can be statically parsed at compiled-time.
```

제가 알기로는 middleware의 matcher에 들어가는 값들은 동적 값이나 스프레드 연산자(...)를 직접 사용할 수 없고, 반드시 값 그 자체 혹은 regex 값으로 들어가야 하는 걸로 알고 있어요. 예를 들면 PUBLIC_ROUTES라는 변수를 선언한 뒤 spread로 가져오는 게 아니라, 다음과 같이 써야할 거에요.

```javascript
export const config = {
  matcher: [
    '/sign-in',
    '/sign-up',
    '/password-reset',
    '/about',
    '/profile',
    '/settings',
  ],
};
```
현재의 변수들은 함수 내에서 검증용으로 사용하므로 삭제하지 말고, config 내의 값들만 바꿔야 할 거 같습니다. 뭔가 컴파일 때 상수로 정해버리는 그런 방법이 있을 거 같은데, 지금은 잘 모르겟네요.. 아시면 댓글 부탁드립니다 @DonghyunKim98

standardwish commented on apps/web/src/service/google-login.service.ts:
```diff
@@ -0,0 +1,63 @@
+    );
+
+    if (!response.ok) {
+      if (response.status === 404) {
```
전 axios만 써봐서 200번대를 제외한 모든 상태 코드가 catch로 가는 줄 알았는데, fetch는 response.ok만 false로 만들고 reject 시키지 않는 사실을 코드보면서 알앗습니다.. API 연결해봐야 알겠지만 로직상 문제는 없어보여요

standardwish reviewed (changes requested):
middleware 부분하고 파일명 변경해주세요~

standardwish reviewed (approved):
문제 없는 거 같습니다! 수고하셨습니다:)

DonghyunKim98 reviewed (approved):
conflict 만 해결해주세요~
